storage/temptable/include/temptable/allocator.h

```diff
@@ -259,8 +261,15 @@ struct Prefer_RAM_over_MMAP_policy_obeying_per_table_limit {
            table_resource_monitor->threshold());
 
     if (table_resource_monitor->consumption() + block_size >
-        table_resource_monitor->threshold())
+        table_resource_monitor->threshold()) {
+      if (table_resource_monitor->consumption() == 0) {
```

Comment:
Here, I agree that there is a logical bug. Having `1M` minimal block allocation contradicts with the fact that `@tmp_table_size` can be set to a lower value.
However, the more important question is why SELECTING from these I_S tables does not automatically switch to on-disk InnoDB temp tables.
According to this
https://dev.mysql.com/doc/refman/8.0/en/internal-temporary-tables.html
this is exactly what should happen
```
When the tmp_table_size limit is reached, MySQL automatically converts the in-memory internal
temporary table to an InnoDB on-disk internal temporary table.
```

Reply:
It does - but at this point, we are in the middle of table creation, which is handled differently. Later, the code that actually fills the table handles this situation.